### PR TITLE
CAL-84 add ability to trim partial lat-lon data

### PR DESCRIPTION
**Pull request by glenhein**
#### What does this PR do?
Add ability to trim partial lat-lon data. This is important in edge cases where one of the data points is missing doing to an error condition in the original KLV data.  
#### Who is reviewing it?
@jrnorth 
@rzwiefel 
@kcwire 
@bdeining 
#### How should this be tested?
Run the unit tests. We don't have a video with the actual edge case.
#### What are the relevant tickets?
CAL-84
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/BaseKlvHandler.java [hunk 1]
```diff
@@ -43,4 +43,10 @@ public String getAttributeName() {
         return Optional.of(new AttributeImpl(getAttributeName(), serials));
     }
 
+    protected <T> int getMinimumListSize(Collection<List<T>> lists) {
+        return lists.stream()
+                .map(List::size)
```
Comment (jrnorth) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/BaseKlvHandler.java]: Alternatively, this could be:
```
.mapToInt(List::size)
.min()
.orElse(0);
```
for some slightly cleaner syntax (imo).

Comment (glenhein) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/BaseKlvHandler.java]: changed

**Comment by bdeining** [comment]
LGTM

**Comment by jrnorth** [comment]
LGTM

File: catalog/video/video-mpegts-stream/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 2]
```diff
@@ -77,7 +77,14 @@
             <property name="stanag4609Processor">
                 <bean class="org.codice.alliance.libs.klv.Stanag4609ProcessorImpl">
                     <argument>
-                        <bean class="org.codice.alliance.libs.klv.OffsetCenterPostProcessor"/>
+                        <bean class="org.codice.alliance.libs.klv.ListPostProcessor">
+                            <argument>
+                                <list>
+                                    <bean class="org.codice.alliance.libs.klv.TrimmingPostProcessor"/>
+                                    <bean class="org.codice.alliance.libs.klv.OffsetCenterPostProcessor"/>
```
Comment (rzwiefel) [inline catalog/video/video-mpegts-stream/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Why not do a reference list of `PostProcessor` here instead of manually instantiating a bean?

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java [hunk 3]
```diff
@@ -156,6 +151,18 @@ public String getLongitude4() {
         return asAttribute(polygonsWkts);
     }
 
+    /**
+     * Trim the arrays of lat and lon values to the same length.
+     */
+    @Override
+    public void trim() {
+        Trimmable.trimList(map.values(), getMinimumListSize());
```
Comment (coyotesqrl) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java]: This functionality concerns me a great deal. I know I have a greater dislike for mutable objects/collections than many other engineers, but you're fiddling with the contents of a collection on the fly in a really gnarly way. Clearly this class is not threadsafe and, I assume, is not intended to be used in any way where it could be shared by threads; however, this is still awfully...bad.

If you were making deep copies and returning immutable collections that would be great; performing surgery on the contents of those collections is just worrisome. I'm confident we can find a better way.

File: libs/klv/src/test/java/org/codice/alliance/libs/klv/TestLatitudeLongitudeHandler.java [hunk 4]
```diff
@@ -32,6 +35,8 @@
 
     private static final String LON = "lon";
 
+    private static final double EPSILON = 0.01;
```
Comment (rzwiefel) [inline libs/klv/src/test/java/org/codice/alliance/libs/klv/TestLatitudeLongitudeHandler.java]: Besides the fact this was only moved to a constant, Is there a reason we're using such a low epsilon? a hundredth of a degree for lat or lon could be something like 1km off

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Trimmable.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     * @param lists    the collection
+     * @param trimSize the desired size of each list
+     * @param <T>      the list type
+     */
+    static <T> void trimList(Collection<List<T>> lists, int trimSize) {
+        lists.forEach(list -> {
+            while (list.size() > trimSize) {
+                list.remove(list.size() - 1);
+            }
+        });
```
Comment (rzwiefel) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Trimmable.java]: Besides the fact this needs to be re-architected, the actual logic could be cleaner with
```java
lists.stream()
                .map(Collection::stream)
                .map(stream -> stream.limit(trimSize))
                .map(stream -> stream.collect(Collectors.toList()))
                .collect(Collectors.toList())
```

**Comment by rzwiefel** [comment]
I think we have some things to address on this before merging.

also we probably should update the original post with who the two committers are (because there's only 1 right now - @kcwire )

File: catalog/video/video-mpegts-stream/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 2 above)
Comment (glenhein) [inline catalog/video/video-mpegts-stream/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I'll have to look at that carefully to make sure the order of the post processors do not matter. I would also want to keep ListPostProcessor because I want my code to only deal with a single PostProcessor at a high level, not a list of processors. At this time, I don't really expect other modules/bundles to register their own post processors.

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java (hunk 3 above)
Comment (glenhein) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java]: The class is modifying it's own internal data, which is really no different if it were to add values to it's internal data. These KlvHandlers build up data is the raw klv data is feed to them. I'd be happy to meet with you and talk about the code.

File: libs/klv/src/test/java/org/codice/alliance/libs/klv/TestLatitudeLongitudeHandler.java (hunk 4 above)
Comment (glenhein) [inline libs/klv/src/test/java/org/codice/alliance/libs/klv/TestLatitudeLongitudeHandler.java]: I could make the epsilon smaller. It's just the fudge factor for comparing floating point numbers.

**Comment by bdeining** [comment]
@jlcsmith

**Comment by jlcsmith** [comment]
LGTM 👍

File: libs/klv/src/test/java/org/codice/alliance/libs/klv/TestGeoBoxHandler.java [hunk 6]
```diff
@@ -140,4 +146,53 @@ public void testAcceptData() throws KlvDecodingException {
...
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LON1, 2));
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LAT2, 3));
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LON2, 4));
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LAT3, 5));
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LON3, 6));
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LAT4, 7));
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LON4, 8));
+
+        // these are the values that should get trimmed
+        geoBoxHandler.accept(KlvUtilities.createTestFloat(LAT1, 1));
```
Comment (jrnorth) [inline libs/klv/src/test/java/org/codice/alliance/libs/klv/TestGeoBoxHandler.java]: If I'm understanding this correctly, this test could be made stronger by using different lat/lon values from the ones you used above in these 4 calls. Since these values are the same as the ones you used above, the assertions below can't prove which value was actually trimmed.

**Comment by bdeining** [comment]
Hero build successful and all unit tests pass

File: libs/klv/src/test/java/org/codice/alliance/libs/klv/TestGeoBoxHandler.java (hunk 6 above)
Comment (glenhein) [inline libs/klv/src/test/java/org/codice/alliance/libs/klv/TestGeoBoxHandler.java]: i'll change the test values to make them distinct

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java [hunk 7]
```diff
@@ -30,7 +29,7 @@
 /**
  * This handler expects four latitude-longitude pairs. It generates a WKT polygon for each four-pair set.
  */
-class GeoBoxHandler extends BaseKlvHandler {
+class GeoBoxHandler extends BaseKlvHandler implements Trimmable {
```
Comment (coyotesqrl) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java]: Is this intended not to be a publicly accessible class?

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/LatitudeLongitudeHandler.java [hunk 8]
```diff
@@ -30,7 +29,7 @@
 /**
  * This handler expects pairs of latitude and longitude values. It generates WKT Points.
  */
-class LatitudeLongitudeHandler extends BaseKlvHandler {
+class LatitudeLongitudeHandler extends BaseKlvHandler implements Trimmable {
```
Comment (coyotesqrl) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/LatitudeLongitudeHandler.java]: Same question re: not public class.

**Comment by coyotesqrl** [comment]
LGTM 👍

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java (hunk 7 above)
Comment (glenhein) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/GeoBoxHandler.java]: hmm, i'll take a look
